### ISSUE TITLE
Adding `!` fails to suppress warning

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#26618]
The sample below shows two problems:
~~1. `T` is actually constrained to a class type, so `return null` should be ok~~
2. Adding `!` did not supress this warning

```C#
public class C {
    public T M<T>()  where T : C {
        return null!; // warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.
    }
}
```

**Comment by cston** [comment]
`!` should suppress the warning (that is a valid bug). But since `T` is constrained to a reference type, the method could be declared as returning `T?` to avoid the warning.

**Comment by jcouv** [comment]
This seems to have been fixed already ([sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAJQkiQDsBjCAFQE8AHFAGhhAEssmATEAagB96cfGwoACCrmTIxAYTEBvALAAoMerFCR46mICyAHmoA+ABQBKdQHcAFhCRjdIOYtUb3YpPARkxZbFgAhADcYgD0YWJWUD5sZADmcgDKABwAbABMAKzOslBkZAD2MBKFZABu9iX+OGJYbDD2UFhiMIV+Zag1uPgQniT2EJR9hQhicJRlyDAIUHEQXK0MffQxUAC2EI0IAHRuGgC+qgdAA===)).
